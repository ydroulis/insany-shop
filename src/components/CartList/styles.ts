"use client"

import styled from 'styled-components'

export const Container = styled.section`
   color: #41414D;

   margin-top: 4.438rem;
`

export const Title = styled.h2`
    font-size: 1.5rem;
    font-weight: 500;

    margin-bottom: 0.5rem;

    text-transform: uppercase;
`

export const Amount = styled.p`
    font-size: 1rem;
    font-weight: 300;

    margin-bottom: 1.438rem;

    span {
        font-weight: 500;
    }
`

export const List = styled.ul`
    list-style: none;

    display: flex;
    flex-direction: column;
    gap: 16px;
`