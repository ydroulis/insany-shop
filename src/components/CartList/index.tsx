import React from 'react';
import CartItem from '../CartItem';

import * as S from './styles';
import { CartProduct } from '@/types/Cart';

interface CartItemProps {
    products: CartProduct[];
}

const CartList: React.FC<CartItemProps> = ({ products }) => {

    const total = products.reduce((acc, product) => acc + product.price, 0);

    const formattedTotal = new Intl.NumberFormat('pt-BR', {
        style: 'currency',
        currency: 'BRL',
    }).format(total);

    return (
        <S.Container>
            <S.Title id="order-summary-title">Resumo do pedido</S.Title>
            <S.Amount
                aria-live="polite"
                aria-labelledby="order-summary-title"
            >
                {`Total (${products.length} ${products.length > 1 ? 'produtos' : 'produto'}) `}
                <span>{formattedTotal}</span>
            </S.Amount>

            <S.List aria-label="Lista de produtos do carrinho">
                {products.map((product, i) => (
                    <li key={i}>
                        <CartItem name={product.name} description={product.description} price={product.price} image={product.image} id={product.id} stock={product.stock} />
                    </li>
                ))}
            </S.List>
        </S.Container>
    );
}

export default CartList;