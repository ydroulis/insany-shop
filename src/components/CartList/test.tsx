import React from 'react';
import { render, screen } from '@testing-library/react';
import CartList from './';

const CartItemMock = ({
    name,
    description,
    price,
    image,
}: {
    name: string;
    description: string;
    price: number;
    image: string;
}) => (
    <div data-testid="cart-item">
        <p>{name}</p>
        <p>{description}</p>
        <p>{price}</p>
        <img src={image} alt={name} />
    </div>
);

CartItemMock.displayName = 'CartItemMock';

jest.mock('../CartItem', () => CartItemMock);

describe('<CartList />', () => {
    const products = [
        {
            id: 1,
            name: 'Produto 1',
            description: 'Desc 1',
            price: 10,
            image: 'img1.png',
            category: 'Roupas',
            stock: 10,
            rating: 4.5,
        },
        {
            id: 2,
            name: 'Produto 2',
            description: 'Desc 2',
            price: 20,
            image: 'img2.png',
            category: 'Roupas',
            stock: 10,
            rating: 4.5,
        },
    ];

    it('should render all elements properly', () => {
        render(<CartList products={products} />);

        const title = screen.getByRole('heading', { name: /resumo do pedido/i });
        expect(title).toBeInTheDocument();

        const totalText = `Total (${products.length} produtos)`;
        expect(
            screen.getByText((content, element) =>
                element?.tagName.toLowerCase() === 'p' && content.includes(totalText)
            )
        ).toBeInTheDocument();
        expect(screen.getByText('R$ 30,00')).toBeInTheDocument();

        const list = screen.getByRole('list', { name: /lista de produtos do carrinho/i });
        expect(list).toBeInTheDocument();

        const items = screen.getAllByTestId('cart-item');
        expect(items.length).toBe(products.length);
        products.forEach((product, index) => {
            expect(items[index]).toHaveTextContent(product.name);
            expect(items[index]).toHaveTextContent(product.description);
            expect(items[index].querySelector('img')).toHaveAttribute('src', product.image);
        });
    });

    it('should match snapshot', () => {
        const { container } = render(<CartList products={products} />);
        expect(container).toMatchSnapshot();
    });
});