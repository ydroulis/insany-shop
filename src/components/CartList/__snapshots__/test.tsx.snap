// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<CartList /> should match snapshot 1`] = `
<div>
  <section
    class="sc-bRKDuR faLdco"
  >
    <h2
      class="sc-hvigdm kLTKGJ"
      id="order-summary-title"
    >
      Resumo do pedido
    </h2>
    <p
      aria-labelledby="order-summary-title"
      aria-live="polite"
      class="sc-fhHczv dzDssg"
    >
      Total (2 produtos) 
      <span>
        R$Â 30,00
      </span>
    </p>
    <ul
      aria-label="Lista de produtos do carrinho"
      class="sc-ggWZvA YqVpg"
    >
      <li>
        <div
          data-testid="cart-item"
        >
          <p>
            Produto 1
          </p>
          <p>
            Desc 1
          </p>
          <p>
            10
          </p>
          <img
            alt="Produto 1"
            src="img1.png"
          />
        </div>
      </li>
      <li>
        <div
          data-testid="cart-item"
        >
          <p>
            Produto 2
          </p>
          <p>
            Desc 2
          </p>
          <p>
            20
          </p>
          <img
            alt="Produto 2"
            src="img2.png"
          />
        </div>
      </li>
    </ul>
  </section>
</div>
`;
